File: ios/native_theme_mode/Sources/native_theme_mode/NativeThemeModePlugin.swift
import Flutter
import UIKit

public class NativeThemeModePlugin: NSObject, FlutterPlugin {
  private static let channelName = "dev.rekan.native_theme_mode"
  private static let suiteName = "native_theme_mode"
  private static let metaStorageKey = "storage_key"
  private static let metaDefaultMode = "default_mode"
  private static let metaPersist = "persist"
  private static let metaEnableIOS = "enable_ios"
  private static let defaultDataKey = "theme_mode"
  private static let modeLight = "light"
  private static let modeDark = "dark"
  private static let modeSystem = "system"
  private static let reservedKeys: Set<String> = [
    metaStorageKey, metaDefaultMode, metaPersist, metaEnableIOS,
  ]

  private let defaults =
    UserDefaults(suiteName: NativeThemeModePlugin.suiteName) ?? .standard

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    let instance = NativeThemeModePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    instance.applyFromDefaults()
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "configure":
      result(configure(arguments: call.arguments))
    case "getThemeMode":
      result(currentMode())
    case "setThemeMode":
      setThemeMode(arguments: call.arguments)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func configure(arguments: Any?) -> String {
    let args = dictionary(from: arguments)
    let storageKey = Self.sanitizeStorageKey(args.stringValue("storageKey") ?? Self.defaultDataKey)
    let defaultMode = Self.sanitizeMode(args.stringValue("defaultMode"))
    let persist = args.boolValue("persist") ?? true
    let enableIOS = args.boolValue("enableIOS") ?? true

    defaults.set(storageKey, forKey: Self.metaStorageKey)
    defaults.set(defaultMode, forKey: Self.metaDefaultMode)
    defaults.set(persist, forKey: Self.metaPersist)
    defaults.set(enableIOS, forKey: Self.metaEnableIOS)

    let mode = currentMode()
    if enableIOS {
      applyUserInterfaceStyle(mode)
    }
    return mode
  }

  private func setThemeMode(arguments: Any?) {
    let args = dictionary(from: arguments)
    let mode = Self.sanitizeMode(args.stringValue("mode"))
    let persist = args.boolValue("persist") ?? defaults.bool(forKey: Self.metaPersist, default: true)
    let enableIOS =
      args.boolValue("enableIOS") ?? defaults.bool(forKey: Self.metaEnableIOS, default: true)

    if persist {
      defaults.set(mode, forKey: dataKey())
    }
    if enableIOS {
      applyUserInterfaceStyle(mode)
    }
  }

  private func applyFromDefaults() {
    let enableIOS = defaults.object(forKey: Self.metaEnableIOS) as? Bool ?? true
    guard enableIOS else { return }
    applyUserInterfaceStyle(currentMode())
  }

  private func applyUserInterfaceStyle(_ mode: String) {
    let style: UIUserInterfaceStyle
    switch mode {
    case Self.modeLight:
      style = .light
    case Self.modeDark:
      style = .dark
    default:
      style = .unspecified
    }

    let apply: () -> Void = {
      for window in self.keyWindows() {
        window.overrideUserInterfaceStyle = style
      }
    }

    if Thread.isMainThread {
      apply()
    } else {
      DispatchQueue.main.async(execute: apply)
    }
  }

  private func currentMode() -> String {
    let defaultMode = Self.sanitizeMode(defaults.string(forKey: Self.metaDefaultMode))
    return Self.sanitizeMode(defaults.string(forKey: dataKey()) ?? defaultMode)
  }

  private func dataKey() -> String {
    Self.sanitizeStorageKey(defaults.string(forKey: Self.metaStorageKey) ?? Self.defaultDataKey)
  }

  private func keyWindows() -> [UIWindow] {
    if #available(iOS 13.0, *) {
      return UIApplication.shared.connectedScenes
        .compactMap { $0 as? UIWindowScene }
        .flatMap(\.windows)
    }
    return UIApplication.shared.windows
  }

  private func dictionary(from arguments: Any?) -> [String: Any] {
    arguments as? [String: Any] ?? [:]
  }

  static func sanitizeStorageKey(_ key: String) -> String {
    if key.isEmpty || reservedKeys.contains(key) {
      return defaultDataKey
    }
    return key
  }

  static func sanitizeMode(_ mode: String?) -> String {
    switch mode {
    case modeLight, modeDark, modeSystem:
      return mode!
    default:
      return modeSystem
    }
  }
}

private extension Dictionary where Key == String, Value == Any {
  func stringValue(_ key: String) -> String? {
    self[key] as? String
  }

  func boolValue(_ key: String) -> Bool? {
    self[key] as? Bool
  }
}

private extension UserDefaults {
  func bool(forKey key: String, default defaultValue: Bool) -> Bool {
    if object(forKey: key) == nil {
      return defaultValue
    }
    return bool(forKey: key)
  }
}

File: android/src/main/kotlin/dev/rekan/native_theme_mode/NativeThemeModePlugin.kt
package dev.rekan.native_theme_mode

import android.app.UiModeManager
import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import androidx.appcompat.app.AppCompatDelegate
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** NativeThemeModePlugin */
class NativeThemeModePlugin :
    FlutterPlugin,
    ActivityAware,
    MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel =
            MethodChannel(
                flutterPluginBinding.binaryMessenger,
                CHANNEL
            )
        channel.setMethodCallHandler(this)
        applyFromPrefs()
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "configure" -> {
                if (!ensureContext(result)) {
                    return
                }
                result.success(configure(call.argumentsMap()))
            }
            "getThemeMode" -> {
                if (!ensureContext(result)) {
                    return
                }
                result.success(currentMode())
            }
            "setThemeMode" -> {
                if (!ensureContext(result)) {
                    return
                }
                setThemeMode(call.argumentsMap())
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        if (::context.isInitialized) {
            applyFromPrefs()
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {}

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}

    override fun onDetachedFromActivity() {}

    private fun ensureContext(result: Result): Boolean {
        if (!::context.isInitialized) {
            result.error("not_attached", "Plugin is not attached to an engine", null)
            return false
        }
        return true
    }

    private fun configure(args: Map<String, Any?>): String {
        val requestedKey = args.stringArg("storageKey") ?: DEFAULT_DATA_KEY
        val storageKey = sanitizeStorageKey(requestedKey)
        val defaultMode = sanitizeMode(args.stringArg("defaultMode"))
        val persist = args.boolArg("persist") ?: true
        val enableAndroid = args.boolArg("enableAndroid") ?: true

        prefs()
            .edit()
            .putString(META_STORAGE_KEY, storageKey)
            .putString(META_DEFAULT_MODE, defaultMode)
            .putBoolean(META_PERSIST, persist)
            .putBoolean(META_ENABLE_ANDROID, enableAndroid)
            .apply()

        val mode = currentMode()
        if (enableAndroid) {
            applyNightMode(mode, persistOs = persist)
        }
        return mode
    }

    private fun setThemeMode(args: Map<String, Any?>) {
        val mode = sanitizeMode(args.stringArg("mode"))
        val persist = args.boolArg("persist") ?: prefs().getBoolean(META_PERSIST, true)
        val enableAndroid =
            args.boolArg("enableAndroid") ?: prefs().getBoolean(META_ENABLE_ANDROID, true)

        if (persist) {
            prefs().edit().putString(dataKey(), mode).commit()
        }
        if (enableAndroid) {
            applyNightMode(mode, persistOs = persist)
        }
    }

    private fun applyFromPrefs() {
        if (!::context.isInitialized) {
            return
        }
        if (!prefs().getBoolean(META_ENABLE_ANDROID, true)) {
            return
        }
        val persist = prefs().getBoolean(META_PERSIST, true)
        applyNightMode(currentMode(), persistOs = persist)
    }

    private fun applyNightMode(
        mode: String,
        persistOs: Boolean
    ) {
        val appCompatMode =
            when (mode) {
                MODE_LIGHT -> AppCompatDelegate.MODE_NIGHT_NO
                MODE_DARK -> AppCompatDelegate.MODE_NIGHT_YES
                else -> AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
            }

        // AppCompat default night mode updates this process. Flutter activities
        // typically list uiMode in configChanges, so this should not recreate
        // the activity.
        AppCompatDelegate.setDefaultNightMode(appCompatMode)

        if (persistOs && Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val uiModeManager = context.getSystemService(UiModeManager::class.java)
            val uiMode =
                when (mode) {
                    MODE_LIGHT -> UiModeManager.MODE_NIGHT_NO
                    MODE_DARK -> UiModeManager.MODE_NIGHT_YES
                    else -> UiModeManager.MODE_NIGHT_AUTO
                }
            uiModeManager.setApplicationNightMode(uiMode)
        }
    }

    private fun currentMode(): String {
        val defaultMode = prefs().getString(META_DEFAULT_MODE, MODE_SYSTEM) ?: MODE_SYSTEM
        return prefs().getString(dataKey(), defaultMode) ?: defaultMode
    }

    private fun dataKey(): String {
        val stored = prefs().getString(META_STORAGE_KEY, DEFAULT_DATA_KEY) ?: DEFAULT_DATA_KEY
        return sanitizeStorageKey(stored)
    }

    private fun prefs(): SharedPreferences {
        return context.getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE)
    }

    companion object {
        private const val CHANNEL = "dev.rekan.native_theme_mode"

        /** Hardcoded plugin preferences file. Host apps never need to open this. */
        private const val PREFS_FILE = "native_theme_mode"

        private const val META_STORAGE_KEY = "storage_key"
        private const val META_DEFAULT_MODE = "default_mode"
        private const val META_PERSIST = "persist"
        private const val META_ENABLE_ANDROID = "enable_android"
        private const val DEFAULT_DATA_KEY = "theme_mode"

        private const val MODE_LIGHT = "light"
        private const val MODE_DARK = "dark"
        private const val MODE_SYSTEM = "system"

        private val RESERVED_KEYS =
            setOf(
                META_STORAGE_KEY,
                META_DEFAULT_MODE,
                META_PERSIST,
                META_ENABLE_ANDROID
            )

        internal fun sanitizeStorageKey(key: String): String {
            return if (key.isEmpty() || key in RESERVED_KEYS) {
                DEFAULT_DATA_KEY
            } else {
                key
            }
        }

        internal fun sanitizeMode(mode: String?): String {
            return when (mode) {
                MODE_LIGHT, MODE_DARK, MODE_SYSTEM -> mode
                else -> MODE_SYSTEM
            }
        }
    }
}

private fun MethodCall.argumentsMap(): Map<String, Any?> {
    val raw = arguments
    if (raw is Map<*, *>) {
        val out = HashMap<String, Any?>()
        for ((key, value) in raw) {
            if (key is String) {
                out[key] = value
            }
        }
        return out
    }
    return emptyMap()
}

private fun Map<String, Any?>.stringArg(key: String): String? {
    val value = this[key]
    return value as? String
}

private fun Map<String, Any?>.boolArg(key: String): Boolean? {
    val value = this[key]
    return value as? Boolean
}
